src/Autofac.Pooling/RegistrationExtensions.cs

```diff
@@ -0,0 +1,369 @@
+                    }
+
+                    // First, we going to create a pooled instance activator, that will be resolved when we want to
+                    // **actually** resolve a new instance (during 'Create').
+                    // The instances themselves are owned by the pool, and will be disposed when the pool disposes
+                    // (or when the instance is not returned to the pool).
+                    var pooledInstanceRegistration = new ComponentRegistration(
+                        Guid.NewGuid(),
+                        instanceActivator,
+                        RootScopeLifetime.Instance,
+                        InstanceSharing.None,
+                        InstanceOwnership.ExternallyOwned,
+                        registration.ResolvePipeline,
+                        new[] { pooledInstanceService },
+                        new Dictionary<string, object?>());
+
+                    registry.Register(pooledInstanceRegistration);
+
+                    var poolService = new PoolService(pooledInstanceRegistration);
+
+                    var poolRegistration = new ComponentRegistration(
+                        Guid.NewGuid(),
+                        new PoolActivator<TLimit>(pooledInstanceService, registrationPolicy),
+                        RootScopeLifetime.Instance,
+                        InstanceSharing.Shared,
+                        InstanceOwnership.OwnedByLifetimeScope,
+                        new[] { poolService },
+                        new Dictionary<string, object?>());
+
+                    registry.Register(poolRegistration);
+
+                    var pooledGetLifetime = tags is null ? CurrentScopeLifetime.Instance : new MatchingScopeLifetime(tags);
+
+                    // Next, create a new registration with a custom activator, that copies metadata and services from
+                    // the original registration. This registration will access the pool and return an instance from it.
+                    var poolGetRegistration = new ComponentRegistration(
+                        Guid.NewGuid(),
+                        new PoolGetActivator<TLimit>(poolService, registrationPolicy),
+                        pooledGetLifetime,
+                        InstanceSharing.Shared,
+                        InstanceOwnership.OwnedByLifetimeScope,
+                        regData.Services,
+                        regData.Metadata);
+
+                    registry.Register(poolGetRegistration);
+
+                    // Finally, add a service pipeline stage to just before the sharing middleware, for each supported service, to extract the pooled instance from the pool instance container.
+                    foreach (var srv in regData.Services)
+                    {
+                        registry.RegisterServiceMiddleware(srv, new PooledInstanceUnpackMiddleware<TLimit>(), MiddlewareInsertionMode.StartOfPhase);
+                    }
+                }
+                else
```

Comment:
It might be more clear to have `if (regData.Lifetime is not PooledLifetime)` at the top and an early `return` to avoid the deep indent/nesting.

Reply:
I've inverted the nesting, but sadly no C#9 `is not` goodies just yet.